README.md
# Lockit CouchDB adapter

[![Build Status](https://travis-ci.org/zemirco/lockit-couchdb-adapter.svg?branch=master)](https://travis-ci.org/zemirco/lockit-couchdb-adapter) [![NPM version](https://badge.fury.io/js/lockit-couchdb-adapter.svg)](http://badge.fury.io/js/lockit-couchdb-adapter)

CouchDB adapter for [Lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-couchdb-adapter`

```js
var adapter = require('lockit-couchdb-adapter');
```

## Configuration

The adapter automatically saves the necessary views to your CouchDB.
You only need the connection string in your `config.js`.

```js
exports.db = 'http://127.0.0.1:5984/';
```

or (long format with custom per-user-db prefix)

```js
exports.db = {
  url: 'http://127.0.0.1:5984/',
  prefix: 'custom/'  // default is 'lockit/',
  database: 'myDB'  // database to write users to, defaults to '_users'
}
```

## Features

### 1. Create user

`adapter.save(name, email, pass, callback)`

 - `name`: String - i.e. 'john'
 - `email`: String - i.e. 'john@email.com'
 - `pass`: String - i.e. 'password123'
 - `callback`: Function - `callback(err, user)` where `user` is the new user now in our database.

The `user` object has the following properties

 - `_id`: unique id from CouchDB
 - `_rev`: revision from CouchDB
 - `password_scheme`: [pbkdf2](http://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2_password_salt_iterations_keylen_callback)
 - `iterations`: pbkdf2 iterations
 - `name`: name chosen during sign up
 - `email`: email that was provided at the beginning
 - `roles`: Array with roles a user has, default is `['user']`
 - `type`: document type for views, default is 'user'
 - `signupToken`: unique token sent to user's email for email verification
 - `signupTimestamp`: Date object to remember when the user signed up
 - `signupTokenExpires`: Date object usually 24h ahead of `signupTimestamp`
 - `failedLoginAttempts`: save failed login attempts during login process, default is `0`
 - `derived_key`: password hash generated by pbkdf2
 - `salt`: salt generated by `crypto.randomBytes()`

```js
adapter.save('john', 'john@email.com', 'secret', function(err, user) {
  if (err) console.log(err);
  console.log(user);
  // {
  //  _id: '8c7cd00c55a25ceb279a8e893d011b3e',
  //  _rev: '1-530a4cd8e67d51daf74e059899c39cd5',
  //  password_scheme: 'pbkdf2',
  //  iterations: 10,
  //  name: 'john',
  //  email: 'john@email.com',
  //  roles: [ 'user' ],
  //  type: 'user',
  //  signupToken: 'fed26ce9-2628-405a-b9fa-285d4a66f4c3',
  //  signupTimestamp: '2013-09-21T10:10:50.357Z',
  //  signupTokenExpires: '2014-01-15T15:27:29.020Z',
  //  failedLoginAttempts: 0,
  //  derived_key: '0a2b1714d6017e7efdc1154ee34c805dea29c06a',
  //  salt: '3a213c87b6a33c70fec767acea697994'
  // }
});
```

### 2. Find user

`adapter.find(match, query, callback)`

 - `match`: String - one of the following: 'name', 'email' or 'signupToken'
 - `query`: String - corresponds to `match`, i.e. 'john@email.com'
 - `callback`:  Function - `callback(err, user)`

```js
adapter.find('name', 'john', function(err, user) {
  if (err) console.log(err);
  console.log(user);
  // {
  //  _id: '8c7cd00c55a25ceb279a8e893d011b3e',
  //  _rev: '1-530a4cd8e67d51daf74e059899c39cd5',
  //  password_scheme: 'pbkdf2',
  //  iterations: 10,
  //  name: 'john',
  //  email: 'john@email.com',
  //  roles: [ 'user' ],
  //  type: 'user',
  //  signupToken: 'fed26ce9-2628-405a-b9fa-285d4a66f4c3',
  //  signupTimestamp: '2013-09-21T10:10:50.357Z',
  //  signupTokenExpires: '2014-01-15T15:27:29.020Z',
  //  failedLoginAttempts: 0,
  //  derived_key: '0a2b1714d6017e7efdc1154ee34c805dea29c06a',
  //  salt: '3a213c87b6a33c70fec767acea697994'
  // }
});
```

### 3. Update user

`adapter.update(user, callback)`

 - `user`: Object - must have `_id` and `_rev` properties
 - `callback`: Function - `callback(err, user)` - `user` is the updated user object

```js
// get a user from db first
adapter.find('name', 'john', function(err, user) {
  if (err) console.log(err);

  // add some new properties to our existing user
  user.newKey = 'and some value';
  user.hasBeenUpdated = true;

  // save updated user to db
  adapter.update(user, function(err, user) {
    if (err) console.log(err);
    // ...
  });
});
```

### 4. Delete user

`adapter.remove(name, callback)`

 - `name`: String
 - `callback`: Function - `callback(err, res)` - `res` is `true` if everything went fine

```js
adapter.remove('john', function(err, res) {
  if (err) console.log(err);
  console.log(res);
  // true
});
```

## Test

`grunt`

## License

Copyright (C) 2013 [Mirco Zeiss](mailto: mirco.zeiss@gmail.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

##### 0.5.1 / 2015-07-03

- update dependencies
- use eslint
- use make for grunt
- fix breaking change (status_code -> statusCode) introduced by `nano`

##### 0.5.0 / 2014-08-18

- make `_users` database name configurable
- update dependencies

##### 0.4.3 / 2014-05-22

- code refactoring
- update dependencies

##### 0.4.2 / 2014-04-28

- initialize all views on startup (fix [#6](https://github.com/zeMirco/lockit-couchdb-adapter/issues/6))
- per-user-db prefix is now configurable (fix [#7](https://github.com/zeMirco/lockit-couchdb-adapter/pull/7))
- update `nano`, `moment`, `bluebird`

##### 0.4.1 / 2014-04-17

- database connection can also be an Object

  ```js
  // this still works - short form
  exports.db = 'http://127.0.0.1:5984/';  

  // this also works - long form (required for MongoDB and SQL)
  exports.db = {
    url: 'http://127.0.0.1:5984/'
  };
  ```

##### 0.4.0 / 2014-04-10

- use optional `request_defaults`
- use CouchDB's native authentication features

  Each user gets an own database named `lockit/[name]`.
  Before 0.4.0 `user` documents were saved within these dbs. Beginning with 0.4.0
  each `user` document is saved in the `_users` db under the name `org.couchdb.user:[name]`.
  Because CouchDB requires the `user` document to have a `name` key I had to change
  the key from `username` to `name`. These changes will be reflected in all the other
  Lockit db adapters.

- add `_security` document to each `lockit/[name]` db

  The per-user-db should only be readable by the rightful owner. So only `john`
  should have access to the `lockit/john` database. CouchDB provides this feature
  out of the box. I simply had to add an extra `_security` document.

  ```js
  {
    members : {
      names : [name]
    }
  }
  ```

##### 0.3.0 / 2014-04-03

- create one database per user - `lockit/:username`
- add `type: 'user'` to user document
- simplify `remove()` method

  ```js
  adapter.remove(match, query, callback)
  ```

  becomes

  ```js
  adapter.remove(username, callback)
  ```

- change database connection string (leave out database name)

  ```js
  exports.db = 'http://127.0.0.1:5984/test';
  ```

  becomes

  ```js
  exports.db = 'http://127.0.0.1:5984/';
  ```

...

##### 0.1.0 / 2014-01-22

 - drop `dbUrl` and use `db` instead
 - use new `config.js` structure

##### 0.0.3 / 2013-12-27

 - improve handling of timestrings by using moment

##### 0.0.2 / 2013-09-21

 - return proper `err` when `delete()` method cannot find a user

##### 0.0.1 / 2013-09-21

 - initial release


test/test.js
'use strict';

var config = require('./config.js');
var should = require('should');

// create db connection with pure nano to test custom prefix
var nano = require('nano')(config.db);

var adapter = require('../index.js')(config);

// wait for connection to db
before(function(done) {
  setTimeout(function() {
    done();
  }, 1000);
});

// start the tests
describe('couchdb adapter for lockit', function() {

  // needed later for test 'find user by signup token'
  var _tmp_signupToken = '';

  it('should create a new user', function(done) {
    adapter.save('john', 'john@email.com', 'secret', function(err, res) {
      if (err) {console.log(err); }
      res.should.have.property('signupToken');
      res.signupToken.should.match(/[0-9a-f]{22}|[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/);
      res.should.have.property('failedLoginAttempts');
      res.failedLoginAttempts.should.equal(0);
      _tmp_signupToken = res.signupToken;
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should create a new user when using an object in config', function(done) {
    var config_alt = JSON.parse(JSON.stringify(config));
    config_alt.db = {
      url: 'http://127.0.0.1:5984/'
    };
    var adapter_alt = require('../index.js')(config_alt);
    adapter_alt.save('jack', 'jack@email.com', 'secret', function(err, res) {
      if (err) {console.log(err); }
      res.name.should.equal('jack');
      done();
    });
  });

  it('should use the custom prefix if provided', function(done) {
    var config_alt = JSON.parse(JSON.stringify(config));
    config_alt.db = {
      url: 'http://127.0.0.1:5984/',
      prefix: 'custom/'
    };
    var adapter_alt = require('../index.js')(config_alt);
    adapter_alt.save('prefix', 'prefix@email.com', 'secret', function(err) {
      if (err) {console.log(err); }

      // make sure db 'custom/prefix' exists
      nano.db.get('custom/prefix', function(getErr, body) {
        if (getErr) {console.log(getErr); }
        body.db_name.should.equal('custom/prefix');
        done();
      });

    });
  });

  it('should find a user by name', function(done) {
    adapter.find('name', 'john', function(err, res) {
      if (err) {console.log(err); }
      res.name.should.equal('john');
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should return null when no user is found', function(done) {
    adapter.find('name', 'jim', function(err, res) {
      if (err) {console.log(err); }
      should.not.exist(err);
      should.not.exist(res);
      done();
    });
  });

  it('should find a user by email', function(done) {
    adapter.find('email', 'john@email.com', function(err, res) {
      if (err) {console.log(err); }
      res.name.should.equal('john');
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should find a user by signup token', function(done) {
    adapter.find('signupToken', _tmp_signupToken, function(err, res) {
      if (err) {console.log(err); }
      res.name.should.equal('john');
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should update an existing user', function(done) {
    adapter.find('name', 'john', function(err, doc) {
      if (err) {console.log(err); }
      doc.test = 'works';
      doc.editet = true;
      adapter.update(doc, function(updateErr, res) {
        if (updateErr) {console.log(updateErr); }
        res.test.should.equal('works');
        res.editet.should.be.true;
        done();
      });
    });
  });

  it('should remove a user', function(done) {
    adapter.save('jeff', 'jeff@email.com', 'secret', function(err) {
      if (err) {console.log(err); }
      adapter.remove('jeff', function(removeErr, res) {
        if (removeErr) {console.log(removeErr); }
        res.should.be.true;
        done();
      });
    });
  });

  it('should return an error when remove cannot find a user', function(done) {
    adapter.remove('steve', function(err) {
      err.message.should.equal('lockit - Cannot find user "steve"');
      done();
    });
  });

});

// remove user and per-user-db
after(function(done) {
  adapter.remove('john', function(err) {
    if (err) {console.log(err); }
    adapter.remove('jack', function(removeErr) {
      if (removeErr) {console.log(removeErr); }

      // use alternative connection for removing user with custom prefix
      var config_alt = JSON.parse(JSON.stringify(config));
      config_alt.db = {
        url: 'http://127.0.0.1:5984/',
        prefix: 'custom/'
      };
      var adapter_alt = require('../index.js')(config_alt);

      adapter_alt.remove('prefix', done);
    });
  });
});


test/config.js

// database url
exports.db = 'http://127.0.0.1:5984/';

// add requests defaults - for example
exports.request_defaults = {
  // proxy: 'http://someproxy'
};

exports.signup = {
  tokenExpiration: '1 day'
};


index.js
'use strict';

var uuid = require('node-uuid');
var ms = require('ms');
var moment = require('moment');
var Bluebird = require('bluebird');
var init = require('./utils/create-views.js');



/**
 * Adapter constructor function.
 *
 * @example
   var Adapter = require('lockit-couchdb-adapter');
   var config = require('./config.js');
   var adapter = new Adapter(config);
 *
 * @param {Object} config - Lockit configuration
 * @constructor
 */
var Adapter = module.exports = function(config) {

  if (!(this instanceof Adapter)) {return new Adapter(config); }

  this.config = config;

  // db connection string
  var url = config.db.url || config.db;
  var usersDbName = config.db.usersDbName || '_users';

  // per-user-db prefix
  this.prefix = config.db.prefix || 'lockit/';

  this.nano = require('nano')({
    url: url,
    request_defaults: config.request_defaults
  });

  // create views
  this._users = this.nano.use(usersDbName);
  init(this._users, function(err) {
    if (err) {throw err; }
  });

};



/**
 * Create a new user.
 *
 * @example
   adapter.save('john', 'john@email.com', 'secret', function(err, user) {
     if (err) console.log(err);
     console.log(user);
     // {
     //  _id: '8c7cd00c55a25ceb279a8e893d011b3e',
     //  _rev: '1-530a4cd8e67d51daf74e059899c39cd5',
     //  password_scheme: 'pbkdf2',
     //  iterations: 10,
     //  name: 'john',
     //  email: 'john@email.com',
     //  roles: [ 'user' ],
     //  type: 'user',
     //  signupToken: 'fed26ce9-2628-405a-b9fa-285d4a66f4c3',
     //  signupTimestamp: '2013-09-21T10:10:50.357Z',
     //  signupTokenExpires: '2014-01-15T15:27:29.020Z',
     //  failedLoginAttempts: 0,
     //  derived_key: '0a2b1714d6017e7efdc1154ee34c805dea29c06a',
     //  salt: '3a213c87b6a33c70fec767acea697994'
     // }
   });
 *
 * @param {String} name - User name
 * @param {String} email - User email
 * @param {String} pw - Plain text user password
 * @param {Function} done - Callback function `function(err, user){}`
 */
Adapter.prototype.save = function(name, email, pw, done) {
  var that = this;

  // create user document in _users db
  var createUser = function() {
    return new Bluebird(function (resolve, reject) {
      // init validation timespan
      var now = moment().toDate();
      var timespan = ms(that.config.signup.tokenExpiration);
      var future = moment().add(timespan, 'ms').toDate();
      // create a new user
      var user = {
        name: name,
        password: pw,
        email: email,
        roles: ['user'],
        type: 'user',
        signupToken: uuid.v4(),
        signupTimestamp: now,
        signupTokenExpires: future,
        failedLoginAttempts: 0
      };
      // add user to db
      that._users.insert(user, 'org.couchdb.user:' + name, function(err, body) {
        if (err) {return reject(err); }
        resolve(body);
      });
    });
  };

  // create per-user-db
  var createDb = function() {
    return new Bluebird(function(resolve, reject) {
      // create new db for user
      var dbName = that.prefix + name;
      that.nano.db.create(dbName, function(err) {
        if (err) {return reject(err); }
        // use new db
        var db = that.nano.use(dbName);
        // create _security document
        var securityDoc = {
           members: {
             names: [name]
           }
        };
        // save security document to db
        db.insert(securityDoc, '_security', function(insertErr, body) {
          if (insertErr) {return reject(insertErr); }
          resolve(body);
        });
      });
    });
  };

  // create both and callback when done
  Bluebird.all([createDb(), createUser()])
  .spread(function(dbResult, userInfo) {
    // get user from db
    that._users.get(userInfo.id, function(err, res) {
      if (err) {return done(err); }
      done(null, res);
    });
  })
  .catch(function(err) {
    done(err);
  });

};



/**
 * Find user. Match is either `'name'`, `'email'` or `'signupToken'`.
 *
 * @example
   adapter.find('name', 'john', function(err, user) {
     if (err) console.log(err);
     console.log(user);
     // {
     //  _id: '8c7cd00c55a25ceb279a8e893d011b3e',
     //  _rev: '1-530a4cd8e67d51daf74e059899c39cd5',
     //  password_scheme: 'pbkdf2',
     //  iterations: 10,
     //  name: 'john',
     //  email: 'john@email.com',
     //  roles: [ 'user' ],
     //  type: 'user',
     //  signupToken: 'fed26ce9-2628-405a-b9fa-285d4a66f4c3',
     //  signupTimestamp: '2013-09-21T10:10:50.357Z',
     //  signupTokenExpires: '2014-01-15T15:27:29.020Z',
     //  failedLoginAttempts: 0,
     //  derived_key: '0a2b1714d6017e7efdc1154ee34c805dea29c06a',
     //  salt: '3a213c87b6a33c70fec767acea697994'
     // }
   });
 *
 * @param {String} match - Property to find user by. `'name'`, `'email'` or `'signupToken'`
 * @param {String} query - Corresponding value to `match`
 * @param {Function} done - Callback function `function(err, user){}`
 */
Adapter.prototype.find = function(match, query, done) {
  var that = this;

  if (match === 'name') {
    // if match is 'name' no need to query the db
    return that._users.get('org.couchdb.user:' + query, function(err, res) {
      if (err && err.statusCode === 404) {return done(null); }
      if (err) {return done(err); }
      // callback
      done(null, res);
    });
  }

  // use a view document and query db
  that._users.view('lockit-user', match, {key: query}, function(err, res) {
    if (err) {return done(err); }
    if (!res.rows.length) {return done(null, null); }
    done(null, res.rows[0].value);
  });

};



/**
 * Update existing user.
 *
 * @example
   // get user from db
   adapter.find('name', 'john', function(err, user) {
     if (err) console.log(err);

     // add some new properties to our existing user
     user.newKey = 'and some value';
     user.hasBeenUpdated = true;

     // save updated user to db
     adapter.update(user, function(err, user) {
       if (err) console.log(err);
       // ...
     });
   });
 *
 * @param {Object} user - Existing user from db
 * @param {Function} done - Callback function `function(err, user){}`
 */
Adapter.prototype.update = function(user, done) {
  var that = this;

  that._users.insert(user, function(err, res) {
    if (err) {return done(err); }
    that._users.get(res.id, done);
  });
};



/**
 * Delete existing user.
 *
 * @example
   adapter.remove('john', function(err, res) {
     if (err) console.log(err);
     console.log(res);
     // true
   });
 *
 * @param {String} name - User name
 * @param {Function} done - Callback function `function(err, res){}`
 */
Adapter.prototype.remove = function(name, done) {
  var that = this;

  // remove user database
  var removeDb = function() {
    return new Bluebird(function(resolve, reject) {
      that.nano.db.destroy(that.prefix + name, function(err, res) {
        if (err && err.statusCode === 404) {
          return reject(new Error('lockit - Cannot find user "' + name + '"'));
        }
        if (err) {return reject(err); }
        resolve(res);
      });
    });
  };

  // and remove user from _users db
  var removeUser = function() {
    return new Bluebird(function (resolve, reject) {
      // get user first
      that._users.get('org.couchdb.user:' + name, function(err, res) {
        if (err && err.statusCode === 404) {
          return reject(new Error('lockit - Cannot find user "' + name + '"'));
        }
        if (err) {return reject(err); }
        // then delete user
        that._users.destroy(res._id, res._rev, function(destroyErr, body) {
          if (destroyErr) {return reject(destroyErr); }
          resolve(body);
        });
      });
    });
  };

  // do both and callback
  Bluebird.all([removeDb(), removeUser()])
  .then(function() {
    done(null, true);
  })
  .catch(function(err) {
    done(err);
  });
};
